include/xtensor/xbuilder.hpp

```diff
@@ -234,6 +349,119 @@ namespace xt
     }
 
     /**
+     * @function identity(std::size_t n, int k = 0)
+     * @brief return identity matrix with 1 on the diagonal. Same as \ref eye(n).
+     */
+    template <class T = bool>
+    inline auto identity(std::size_t n)
+    {
+        return eye<T>({n, n});
+    }
+
+    /**
+     * @function diagonal(const xexpression<T>& arr)
+     * @brief Returns the elements on the diagonal of arr
+     *
+     * @param arr the input array
+     *
+     * @return xexpression with values of the diagonal
+     */
+    template <class T>
+    inline auto diagonal(const xexpression<T>& arr)
+    {
+        const T& arr_dc = arr.derived_cast();
+        return detail::make_xgenerator(detail::fn_impl<detail::diagonal_fn<T>>(detail::diagonal_fn<T>(arr_dc)), {arr_dc.shape()[0]});
+    }
+
+    /**
+     * @function diag(const xexpression<T>& arr)
+     * @brief xexpression with values of arr on the diagonal, zeroes otherwise
+     *
+     * @param arr the 1D input array of length n
+     *
+     * @return xexpression function with shape n x n and arr on the diagonal
+     */
+    template <class T>
+    inline auto diag(const xexpression<T>& arr)
+    {
+        const T& arr_dc = arr.derived_cast();
+        return detail::make_xgenerator(detail::fn_impl<detail::diag_fn<T>>(detail::diag_fn<T>(arr_dc)), 
+                                       {arr_dc.shape()[0], arr_dc.shape()[0]});
+    }
+
+    /**
+     * @function fliplr(const xexpression<T>& arr)
+     * @brief Flip xexpression in the left/right direction. Essentially flips the last axis.
+     *
+     * @param arr the input array
+     *
+     * @return xexpression with values flipped in left/right direction
+     */
+    template <class T>
+    inline auto fliplr(const xexpression<T>& arr)
+    {
+        const T& arr_dc = arr.derived_cast();
+        return detail::make_xgenerator(detail::fn_impl<detail::fliplr_fn<T>>(detail::fliplr_fn<T>(arr_dc)), 
+                                       arr_dc.shape());
+    }
+
+    /**
+     * @function flipud(const xexpression<T>& arr)
+     * @brief Flip xexpression in the up/down direction. Essentially flips the last axis.
+     *
+     * @param arr the input array
+     *
+     * @return xexpression with values flipped in up/down direction
+     */
+    template <class T>
+    inline auto flipud(const xexpression<T>& arr)
+    {
+        const T& arr_dc = arr.derived_cast();
+        return detail::make_xgenerator(detail::fn_impl<detail::flipud_fn<T>>(detail::flipud_fn<T>(arr_dc)), 
```

Comment:
Regarding the remarks in ``flipup_fn`` implementation, this line would become:
``return detail::make_xgenerator(detail::flipud_fn<T>(arr_dc)) ... ``